File: leetcode/1288.删除被覆盖区间/1288.删除被覆盖区间.py
from typing import List


class Solution:
    def removeCoveredIntervals(self, intervals: List[List[int]]) -> int:
        # 按照起点升序排序，起点相同时终点降序
        intvs = sorted(intervals, key=lambda x: [x[0], -x[1]])
        # 记录合并区间的起点和终点
        left, right = intvs[0][0], intvs[0][1]
        res = 0
        for lf, r in intvs[1:]:
            # 找到覆盖区间
            if left <= lf and right >= r:
                res += 1
            # 找到相交区间，合并
            if lf <= right <= r:
                right = r
            # 完全不相交，更新起点和终点
            if right < lf:
                left = lf
                right = r
        return len(intervals) - res


if __name__ == '__main__':
    # intervals = [[1, 4], [3, 6], [2, 8]]
    intervals = [[1, 4], [2, 3]]
    result = Solution().removeCoveredIntervals(intervals)
    print(result)

File: leetcode/1288.删除被覆盖区间/1288.删除被覆盖区间.go
package main

import (
	"sort"
)

func removeCoveredIntervals(intervals [][]int) int {
	// 排序规则，左端从大到小，右端从小到大
	less := func(i, j int) bool {
		if intervals[i][0] == intervals[j][0] {
			return intervals[i][1] >= intervals[j][1]
		}
		return intervals[i][0] <= intervals[j][0]
	}
	// 进行排序
	sort.Slice(intervals, less)
	// 初始化记录区间
	left := intervals[0][0]
	right := intervals[0][1]
	res := 0
	n := len(intervals)
	for i := 1; i < n; i++ {
		// 记录左端小于当前左，大于当前右，找到覆盖区，数量+1
		if left <= intervals[i][0] && right >= intervals[i][1] {
			res += 1
		}
		// 记录右端在当前左右之间，找到交叉区间，更新记录右端位置为当前右
		if right >= intervals[i][0] &&  right <= intervals[i][1] {
			right = intervals[i][1]
		}
		// 记录右端位置小于当前左端，更新记录区间左右为当前左右
		if right <= intervals[i][0]{
			left = intervals[i][0]
			right = intervals[i][1]
		}
	}
	// 长度-覆盖数量 = 剩余数目
	return n - res
}

func main() {
	intervals := [][]int{{1, 4}, {3, 6}, {2, 8}}
	result := removeCoveredIntervals(intervals)
	println(result)
}
